### PR TITLE
Trailing throttle syncs to front-end

### PR DESCRIPTION
**Pull request by pankgeorg**
I'm using a slightly modified version of `Flux.jl`'s throttle: https://github.com/FluxML/Flux.jl/blob/8afedcd6723112ff611555e350a8c84f4e1ad686/src/utils.jl#L662

### CODE REVIEW

**Comment by pankgeorg** [comment]
Converting this to draft as I would like to check whether we can add a lock to the function call; I don't trust concurrency 🎈✨⭐

**Comment by pankgeorg** [comment]
Added 'thread safety', which means that the function _cannot run two times simultaneously_.
The caveat is that (in tightly called sequences) the timeout time gets extended by the time the function is running.

The last thing to consider is, do mutations that may happen 'simultaneously' matter? should every throttled function call get its own copy of the state??

**Comment by fonsp** [comment]
In the first paragraph of your comment, aren't you saying that the second paragraph is not possible?

**Comment by fonsp** [comment]
Can you add some super basic tests with `sleep` to make sure that the basic throttling is working? The code got a bit too complicated to just review it

**Comment by pankgeorg** [comment]
> In the first paragraph of your comment, aren't you saying that the second paragraph is not possible?

Not exactly - I'm saying that if `f` is throttled by the function, and has `argument1`, `argument2` as arguments, then, `f` will never run when `f` runs. (first paragraph)

BUT if `g` is mutating `argument1`, `argument2` there is no guarantee that `f` and `g` don't run such as they corrupt each other's data. (second paragraph)

I'm not even sure if that's possible by Julia's thread model (since our `f` doesn't yield, but it may do some network stuff that counts as 'idle'??)

The thing is that we'll probably need to be locking the resource (`argument1`, `argument2` = `session`, `notebook`).

**Comment by fonsp** [comment]
Why does `f` take arguments?

File: src/evaluation/Run.jl [hunk 1]
```diff
@@ -47,7 +47,8 @@ function run_reactive!(session::ServerSession, notebook::Notebook, old_topology:
 	end
 
 	# Send intermediate updates to the clients at most 20 times / second during a reactive run. (The effective speed of a slider is still unbounded, because the last update is not throttled.)
-	send_notebook_changes_throttled = throttled(1.0/20, 0.0/20) do
+	# flush_send_notebook_changes_throttled, 
+	send_notebook_changes_throttled, flush_notebook_changes = throttle(1.0/20; leading=false, trailing=true) do
 		send_notebook_changes!(ClientRequest(session=session, notebook=notebook))
```
Comment (pankgeorg) [inline src/evaluation/Run.jl:52]: This code will run at most once per 1/20 sec, `@async`. 
BUT there is no guarantee I can think of that [X] code will not run right before/right after/inbetween

**Comment by fonsp** [comment]
>  locking the resource (argument1, argument2 = session, notebook).

We already lock `notebook`:

https://github.com/fonsp/Pluto.jl/blob/7005c6bb8d236e297962386e3e5d53ed4c9b378e/src/evaluation/Run.jl#L10

**Comment by pankgeorg** [comment]
> Why does `f` take arguments? It makes more sense to drop that feature

The feature is not related; the `() -> send_notebook_changes!(ClientRequest(session=session, notebook=notebook))` holds references to `session` and `notebook`. 

> > locking the resource (argument1, argument2 = session, notebook).
> 
> We already lock `notebook`, right?
> 
> https://github.com/fonsp/Pluto.jl/blob/7005c6bb8d236e297962386e3e5d53ed4c9b378e/src/evaluation/Run.jl#L10

all these functions all hold the lock, I think.

**Comment by fonsp** [comment]
> The feature is not related

I see, I can try to think about the thread safe stuff tomorrow with a CLEAR brain :) It does seem like the code would be simpler if you deleted the args, kwargs, and result features. `later` could just be a boolean, for example.

**Pull request merged**